src/util/invariant.h

```diff
@@ -72,41 +72,43 @@ Author: Martin Brain, martin.brain@diffblue.com
 /// family of macros, allowing constructs like
 /// `INVARIANT(x==y, my_invariantt, (T1)actual1, (T2)actual2, ...)`
 ///
-class invariant_failedt: public std::logic_error
```

Comment:
I recall someone being very emphatically passionate that this had to inherit from one of the standard exception handlers.  i don't now recall why.  Does anyone know?  Why is it better to not do so.

Reply:
The idea behind it is that we don't want a generic exception catcher

```
catch (exception &e)
```

to catch these exceptions if they propagate. Now an ellipsis (catch all - `catch (...)`) will still catch it, which we don't think is a good idea, but since we are not entirely sure of how further up the callstack these exceptions can propagate and where exactly (pathwise) these catchers are, we are going to fix this issue as a later part of this work.

Reply:
OK... at the risk of being a stuck record on this; don't over-estimate what you can do after catching an exception.  If an invariant has failed then there is at least one significant, relevant bug in CPROVER.  After that data structures should not be trusted and results should especially not be trusted.  So... it might be that the only sensible thing to do is the same as the handlers for ```catch (exception &e)``` or ```catch (...)```.  YMMV.

Reply:
In this case, it's still my opinion that there should be a special handler that handles this type, even if it duplicates the body of the catch all handlers, even if it's only for documentation purposes - making it clear that this is desired behaviour. As it is right now, the behaviour, even if correct or the only logical one, appears to be accidental.

Reply:
Up to you.  As I said to start with, at one point, I recall someone being very passionate about this but I can't remember why.  As long as there is a good reason; it's all good.